File: chatdownloader/src/metrics/copypastaleader.py
"""
The copypasta leader metric.

TODO: This guy is the reason why I can't do multi-processing just
yet. Need a merge() function at some point to consolidate data between
multiple processes somehow.

TODO: To fix this, I probably need to perform a union-find
TODO: To fix this^2, I think we might need to assume transitivity. Also it seems likely that I have to merge state or something, looks super troublesome smh my head
"""

import heapq
import logging
import itertools
from _types import Comment

from .abstractmetric import AbstractMetric

WEIGHT_COPYPASTA = 0.3
CHAIN_GRACE = 10
MATCHING_THRESHOLD = 0.6

class CopypastaLeader(AbstractMetric):
    """
    The Copypasta leader metric.

    TODO: You probably need to rewrite this
    """
    def __init__(self):
        self.__heap = []

    @classmethod
    def can_parallelize(cls) -> bool:
        return False

    @staticmethod
    def _lcs(lhs: str, rhs: str) -> int:
        # the dp approach, copied from somewhere
        m = len(lhs)
        n = len(rhs)

        L = [[None] * (n + 1) for i in range(m + 1)]

        for i in range(m + 1):
            for j in range(n + 1):
                if i == 0 or j == 0:
                    L[i][j] = 0
                elif lhs[i - 1] == rhs[j - 1]:
                    L[i][j] = L[i - 1][j - 1] + 1
                else:
                    L[i][j] = max(L[i - 1][j], L[i][j - 1])

        return L[m][n]

    @staticmethod
    def padTo(target: str, padding: str, maxlen: int) -> str:
        return (target + (padding *
                          ((maxlen - len(target)) // len(padding) + 1))
                )[:maxlen]

    @classmethod
    def get_name(cls):
        return 'copypasta'

    def get_metric(self, comment: Comment,
                   sequence_no: int) -> list[dict[str, int]]:
        text = ' '.join(
            fragment.text for fragment in comment.message.fragments)

        if len(text) == 0:
            return []

        # if empty heap then just go for it
        if len(self.__heap) == 0:
            heapq.heappush(self.__heap, (sequence_no, text,
                                         comment.commenter._id,
                                         sequence_no))

        logging.debug('Size of heap: %d', len(self.__heap))

        # go through everything in the heap and find the best matching string
        matching_scores = [self._lcs(
            self.padTo(item[1], item[1],
                       len(text)), text) // len(text)
                           for item in self.__heap]
        best_match = max(enumerate(self.__heap),
                         key=lambda item: matching_scores[item[0]])

        if matching_scores[best_match[0]] < MATCHING_THRESHOLD:
            # tuple: (last seq number, text, id, original seq number )
            heapq.heappush(self.__heap, (sequence_no, text,
                                         comment.commenter._id,
                                         sequence_no))
        else:
            # in-place updating of the tuple.
            # TODO: find a better way to do this, i'm too lazy atm
            item = self.__heap[best_match[0]]
            self.__heap[best_match[0]] = (
                sequence_no,
                *item[1:]
            )
            heapq.heapify(self.__heap)

        # evict old heap top
        result = {}
        logging.debug('First item on head has seq no: %d', self.__heap[0][0])
        logging.debug('Last item on head has seq no: %d', self.__heap[-1][0])
        while len(self.__heap) > 0 and (
                sequence_no - self.__heap[0][0]) > CHAIN_GRACE:
            item = heapq.heappop(self.__heap)
            result[item[2]] = (item[0] - item[3]) * WEIGHT_COPYPASTA

        return result

    def finish(self):
        return {item[2]: (item[0] - item[3]) * WEIGHT_COPYPASTA
                for item in self.__heap}

File: chatdownloader/src/metrics/copypastaleader.rs
use log::debug;

use crate::_types::twitchtypes::Comment;
use crate::metrics::metrictrait::AbstractMetric;

use std::collections::HashMap;

const WEIGHT_COPYPASTA: f32 = 0.3;
const CHAIN_GRACE: u32 = 10;
const MATCHING_THRESHOLD: f32 = 0.6;

#[derive(Default, Debug)]
pub struct CopypastaLeader {
    history: Vec<(u32, String, String, u32)>,
}

impl AbstractMetric for CopypastaLeader {
    async fn new() -> Self {
        Self {
            history: Vec::new(),
        }
    }
    
    fn can_parallelize(&self) -> bool {
        false
    }

    fn get_name(&self) -> String {
        String::from("copypasta")
    }

    fn get_metric(&mut self, comment: Comment, sequence_no: u32) -> HashMap<String, f32> {
        let text = comment.message.fragments.iter()
            .map(|fragment| fragment.text.clone())
            .collect::<Vec<String>>()
            .join(" ");

        if text.is_empty() {
            return HashMap::new();
        }

        // Evaluate or initialize the list
        if self.history.is_empty() {
            self.history.push((sequence_no, text.clone(), comment.commenter._id.clone(), sequence_no));
        }

        debug!("Size of heap: {}", self.history.len());

        // Find the best matching string in the list 
        let mut best_match = None;
        let mut best_match_score = 0.0;
        for item in self.history.iter() {
            let lcs = Self::lcs(&text, &item.1);
            let score = lcs.len() as f32 / text.len().max(item.1.len()) as f32;
            if score > best_match_score {
                best_match = Some(item);
                best_match_score = score;
            }
        }

        // If the best match is above the threshold, update the list
        if best_match_score > MATCHING_THRESHOLD {
            let item = best_match.unwrap();
            self.history.push((sequence_no, text.clone(), comment.commenter._id.clone(), item.3));
        } else {
            self.history.push((sequence_no, text.clone(), comment.commenter._id.clone(), sequence_no));
        }

        // Sort the list
        self.history.sort_by_key(|item| item.0);
        debug!("Items on heap: {:?}", self.history.iter().map(|item| item.0).collect::<Vec<u32>>());

        debug!("First item on heap has seq no: {:?}", self.history.first().unwrap().0);
        debug!("Last item on heap has seq no: {:?}", self.history.last().unwrap().0);

        // Evict old list top
        let mut result = HashMap::new();
        while self.history.len() > 0 && (sequence_no - self.history[0].0) > CHAIN_GRACE {
            let item = self.history.remove(0);
            result.insert(item.2.clone(), (item.0 - item.3) as f32 * WEIGHT_COPYPASTA);
        }

        result
    }

    fn finish(&self) -> HashMap<String, f32> {
        self.finish()
    }
}

impl CopypastaLeader {
    fn lcs(s1: &str, s2: &str) -> String {
        let mut max: Option<String> = None; // Holds value of string with maximum length
        let mut current = String::new(); // String container to hold current longest value
        let mut s1_iter = s1.chars().peekable(); // Peekable iterator for string s1
        let mut s2_iter = s2.chars(); //Iterator for string s2
        let mut s2_prev_pos = s2_iter.clone(); // Iterator that holds position of previous location of first iterator
        let mut s1_prev_pos = s1_iter.clone(); // Peekable iterator used to make sure all possible combinations are located.
    
        loop {
            let s1_char = s1_iter.next(); // Get character in s1
    
            if current.is_empty() {
                // If no consecutive string found yet store location of iterator
                s1_prev_pos = s1_iter.clone()
            }
    
            match s1_char {
                Some(s1_char) => loop {
                    match s2_iter.next() {
                        Some(s2_char) if s1_char == s2_char => {
                            current.push(s1_char);
                            s2_prev_pos = s2_iter.clone();
                            break;
                        }
                        Some(_) => continue,
                        None => {
                            s2_iter = s2_prev_pos.clone();
                            break;
                        }
                    }
                },
                None => match s1_prev_pos.peek() {
                    Some(_) => {
                        if max.as_ref().map_or(true, |s| s.len() < current.len()) {
                            max = Some(current.clone());
                        }
                        current.clear();
    
                        s1_iter = s1_prev_pos.clone();
                        s2_iter = s2.chars();
                    }
                    None => break,
                },
            }
        }
    
        max.unwrap_or_default()
    }

    pub fn finish(&self) -> HashMap<String, f32> {
        self.history.iter()
            .map(|item| (item.2.clone(), (item.0 - item.3) as f32 * WEIGHT_COPYPASTA))
            .collect()
    }
}
